File: enterprise/cloud/gcp/cloud-functions/package-registry-function.py
"""
TuskLang Package Registry Google Cloud Function
Serverless package management for TuskLang on GCP
"""

import json
import logging
import os
from datetime import datetime, timedelta
from typing import Dict, Any, Optional
import base64
import hashlib

from google.cloud import storage
from google.cloud import firestore
from google.cloud import pubsub_v1
from google.cloud import secretmanager
import functions_framework

# Configure logging
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

# Initialize GCP clients
storage_client = storage.Client()
firestore_client = firestore.Client()
publisher = pubsub_v1.PublisherClient()
secret_client = secretmanager.SecretManagerServiceClient()

# Configuration
BUCKET_NAME = os.environ.get('PACKAGE_BUCKET_NAME', 'tusklang-packages')
COLLECTION_NAME = os.environ.get('PACKAGE_COLLECTION_NAME', 'packages')
TOPIC_NAME = os.environ.get('PUBSUB_TOPIC_NAME', 'package-events')
PROJECT_ID = os.environ.get('GCP_PROJECT_ID')

class PackageRegistryGCP:
    def __init__(self):
        self.bucket = storage_client.bucket(BUCKET_NAME)
        self.collection = firestore_client.collection(COLLECTION_NAME)
        self.topic_path = publisher.topic_path(PROJECT_ID, TOPIC_NAME)
        
    def upload_package(self, request_data: Dict[str, Any]) -> Dict[str, Any]:
        """Handle package upload requests"""
        try:
            # Parse request
            package_name = request_data.get('name')
            package_version = request_data.get('version')
            package_data = request_data.get('data')
            package_metadata = request_data.get('metadata', {})
            
            if not all([package_name, package_version, package_data]):
                return self._error_response(400, "Missing required fields")
            
            # Validate package
            validation_result = self._validate_package(package_name, package_version, package_data)
            if not validation_result['valid']:
                return self._error_response(400, validation_result['message'])
            
            # Generate package ID and blob name
            package_id = f"{package_name}-{package_version}"
            blob_name = f"packages/{package_name}/{package_version}/package.tsk"
            
            # Check if package already exists
            if self._package_exists(package_name, package_version):
                return self._error_response(409, "Package version already exists")
            
            # Upload to Cloud Storage
            blob = self.bucket.blob(blob_name)
            package_bytes = base64.b64decode(package_data)
            
            blob.metadata = {
                'package-name': package_name,
                'package-version': package_version,
                'upload-time': datetime.utcnow().isoformat()
            }
            
            blob.upload_from_string(
                package_bytes,
                content_type='application/octet-stream'
            )
            
            # Store metadata in Firestore
            package_doc = {
                'package_id': package_id,
                'name': package_name,
                'version': package_version,
                'blob_name': blob_name,
                'size': len(package_bytes),
                'metadata': package_metadata,
                'upload_time': datetime.utcnow(),
                'download_count': 0,
                'last_download': None,
                'checksum': hashlib.sha256(package_bytes).hexdigest()
            }
            
            self.collection.document(package_id).set(package_doc)
            
            # Publish event to Pub/Sub
            self._publish_event('package.uploaded', {
                'package_id': package_id,
                'name': package_name,
                'version': package_version,
                'size': len(package_bytes)
            })
            
            return self._success_response({
                'package_id': package_id,
                'download_url': f"https://storage.googleapis.com/{BUCKET_NAME}/{blob_name}",
                'message': 'Package uploaded successfully'
            })
            
        except Exception as e:
            logger.error(f"Upload error: {str(e)}")
            return self._error_response(500, f"Upload failed: {str(e)}")
    
    def download_package(self, package_name: str, package_version: str) -> Dict[str, Any]:
        """Handle package download requests"""
        try:
            if not all([package_name, package_version]):
                return self._error_response(400, "Missing package name or version")
            
            package_id = f"{package_name}-{package_version}"
            
            # Get package metadata from Firestore
            doc_ref = self.collection.document(package_id)
            doc = doc_ref.get()
            
            if not doc.exists:
                return self._error_response(404, "Package not found")
            
            package_data = doc.to_dict()
            
            # Generate signed URL for download
            blob = self.bucket.blob(package_data['blob_name'])
            signed_url = blob.generate_signed_url(
                version="v4",
                expiration=timedelta(hours=1),
                method="GET"
            )
            
            # Update download statistics
            self._update_download_stats(package_id)
            
            # Publish download event
            self._publish_event('package.downloaded', {
                'package_id': package_id,
                'name': package_name,
                'version': package_version
            })
            
            return self._success_response({
                'package_id': package_id,
                'download_url': signed_url,
                'metadata': package_data.get('metadata', {}),
                'size': package_data['size']
            })
            
        except Exception as e:
            logger.error(f"Download error: {str(e)}")
            return self._error_response(500, f"Download failed: {str(e)}")
    
    def search_packages(self, search_term: str = "", limit: int = 20, offset: int = 0) -> Dict[str, Any]:
        """Handle package search requests"""
        try:
            # Build Firestore query
            query = self.collection
            
            if search_term:
                # Firestore doesn't support full-text search, so we'll do a simple contains check
                # In production, you'd use a search service like Algolia or Elasticsearch
                query = query.where('name', '>=', search_term).where('name', '<=', search_term + '\uf8ff')
            
            # Execute query
            docs = query.limit(limit).offset(offset).stream()
            
            # Process results
            packages = []
            for doc in docs:
                data = doc.to_dict()
                packages.append({
                    'package_id': data['package_id'],
                    'name': data['name'],
                    'version': data['version'],
                    'download_count': data.get('download_count', 0),
                    'upload_time': data['upload_time'].isoformat(),
                    'metadata': data.get('metadata', {})
                })
            
            return self._success_response({
                'packages': packages,
                'total': len(packages),
                'next_offset': offset + limit if len(packages) == limit else None
            })
            
        except Exception as e:
            logger.error(f"Search error: {str(e)}")
            return self._error_response(500, f"Search failed: {str(e)}")
    
    def get_package_info(self, package_name: str, package_version: str) -> Dict[str, Any]:
        """Get detailed package information"""
        try:
            if not all([package_name, package_version]):
                return self._error_response(400, "Missing package name or version")
            
            package_id = f"{package_name}-{package_version}"
            
            # Get package metadata from Firestore
            doc_ref = self.collection.document(package_id)
            doc = doc_ref.get()
            
            if not doc.exists:
                return self._error_response(404, "Package not found")
            
            package_data = doc.to_dict()
            download_url = f"https://storage.googleapis.com/{BUCKET_NAME}/{package_data['blob_name']}"
            
            return self._success_response({
                'package_id': package_id,
                'name': package_data['name'],
                'version': package_data['version'],
                'download_url': download_url,
                'size': package_data['size'],
                'download_count': package_data.get('download_count', 0),
                'upload_time': package_data['upload_time'].isoformat(),
                'last_download': package_data.get('last_download', {}).isoformat() if package_data.get('last_download') else None,
                'metadata': package_data.get('metadata', {}),
                'checksum': package_data.get('checksum')
            })
            
        except Exception as e:
            logger.error(f"Package info error: {str(e)}")
            return self._error_response(500, f"Failed to get package info: {str(e)}")
    
    def _validate_package(self, name: str, version: str, data: str) -> Dict[str, Any]:
        """Validate package data"""
        try:
            # Basic validation
            if not name or len(name) > 100:
                return {'valid': False, 'message': 'Invalid package name'}
            
            if not version or len(version) > 50:
                return {'valid': False, 'message': 'Invalid package version'}
            
            if not data:
                return {'valid': False, 'message': 'Package data is empty'}
            
            # Check size limit
            package_bytes = base64.b64decode(data)
            if len(package_bytes) > 50 * 1024 * 1024:  # 50MB limit
                return {'valid': False, 'message': 'Package too large'}
            
            # Check for valid TuskLang syntax (basic check)
            package_text = package_bytes.decode('utf-8', errors='ignore')
            if not package_text.strip().startswith('#') and ':' not in package_text:
                return {'valid': False, 'message': 'Invalid TuskLang format'}
            
            return {'valid': True, 'message': 'Package is valid'}
            
        except Exception as e:
            return {'valid': False, 'message': f'Validation error: {str(e)}'}
    
    def _package_exists(self, name: str, version: str) -> bool:
        """Check if package already exists"""
        try:
            package_id = f"{name}-{version}"
            doc_ref = self.collection.document(package_id)
            doc = doc_ref.get()
            return doc.exists
        except Exception:
            return False
    
    def _update_download_stats(self, package_id: str):
        """Update download statistics"""
        try:
            doc_ref = self.collection.document(package_id)
            doc_ref.update({
                'download_count': firestore.Increment(1),
                'last_download': datetime.utcnow()
            })
        except Exception as e:
            logger.error(f"Failed to update download stats: {str(e)}")
    
    def _publish_event(self, event_type: str, event_data: Dict[str, Any]):
        """Publish event to Pub/Sub"""
        try:
            message = {
                'event_type': event_type,
                'timestamp': datetime.utcnow().isoformat(),
                'data': event_data
            }
            
            future = publisher.publish(
                self.topic_path,
                json.dumps(message).encode('utf-8')
            )
            future.result()  # Wait for the publish to complete
            
        except Exception as e:
            logger.error(f"Failed to publish event: {str(e)}")
    
    def _success_response(self, data: Dict[str, Any]) -> Dict[str, Any]:
        """Generate success response"""
        return {
            'statusCode': 200,
            'headers': {
                'Content-Type': 'application/json',
                'Access-Control-Allow-Origin': '*',
                'Access-Control-Allow-Headers': 'Content-Type',
                'Access-Control-Allow-Methods': 'GET, POST, PUT, DELETE, OPTIONS'
            },
            'body': json.dumps(data)
        }
    
    def _error_response(self, status_code: int, message: str) -> Dict[str, Any]:
        """Generate error response"""
        return {
            'statusCode': status_code,
            'headers': {
                'Content-Type': 'application/json',
                'Access-Control-Allow-Origin': '*',
                'Access-Control-Allow-Headers': 'Content-Type',
                'Access-Control-Allow-Methods': 'GET, POST, PUT, DELETE, OPTIONS'
            },
            'body': json.dumps({
                'error': message,
                'status_code': status_code
            })
        }

# Initialize handler
handler = PackageRegistryGCP()

@functions_framework.http
def package_registry(request):
    """Main Cloud Function handler"""
    try:
        # Handle CORS preflight requests
        if request.method == 'OPTIONS':
            return ('', 204, {
                'Access-Control-Allow-Origin': '*',
                'Access-Control-Allow-Headers': 'Content-Type',
                'Access-Control-Allow-Methods': 'GET, POST, PUT, DELETE, OPTIONS'
            })
        
        # Route requests based on HTTP method and path
        if request.method == 'POST' and request.path == '/packages':
            request_data = request.get_json()
            return handler.upload_package(request_data)
            
        elif request.method == 'GET' and '/packages/' in request.path and '/download' in request.path:
            # Extract package name and version from path
            path_parts = request.path.split('/')
            if len(path_parts) >= 4:
                package_name = path_parts[2]
                package_version = path_parts[3]
                return handler.download_package(package_name, package_version)
            else:
                return handler._error_response(400, "Invalid path")
                
        elif request.method == 'GET' and request.path == '/packages/search':
            search_term = request.args.get('q', '')
            limit = int(request.args.get('limit', 20))
            offset = int(request.args.get('offset', 0))
            return handler.search_packages(search_term, limit, offset)
            
        elif request.method == 'GET' and '/packages/' in request.path:
            # Extract package name and version from path
            path_parts = request.path.split('/')
            if len(path_parts) >= 4:
                package_name = path_parts[2]
                package_version = path_parts[3]
                return handler.get_package_info(package_name, package_version)
            else:
                return handler._error_response(400, "Invalid path")
        else:
            return handler._error_response(404, "Endpoint not found")
            
    except Exception as e:
        logger.error(f"Cloud Function error: {str(e)}")
        return handler._error_response(500, f"Internal server error: {str(e)}") 

File: enterprise/cloud/azure/function-app/package-registry-function.cs
using System;
using System.IO;
using System.Threading.Tasks;
using Microsoft.AspNetCore.Mvc;
using Microsoft.Azure.WebJobs;
using Microsoft.Azure.WebJobs.Extensions.Http;
using Microsoft.AspNetCore.Http;
using Microsoft.Extensions.Logging;
using Newtonsoft.Json;
using Azure.Storage.Blobs;
using Azure.Storage.Blobs.Models;
using Azure.Data.Tables;
using System.Text;
using System.Collections.Generic;
using System.Linq;

namespace TuskLang.PackageRegistry
{
    public class PackageRegistryFunction
    {
        private readonly BlobServiceClient _blobServiceClient;
        private readonly TableClient _tableClient;
        private readonly ILogger<PackageRegistryFunction> _logger;

        public PackageRegistryFunction(
            BlobServiceClient blobServiceClient,
            TableClient tableClient,
            ILogger<PackageRegistryFunction> logger)
        {
            _blobServiceClient = blobServiceClient;
            _tableClient = tableClient;
            _logger = logger;
        }

        [FunctionName("UploadPackage")]
        public async Task<IActionResult> UploadPackage(
            [HttpTrigger(AuthorizationLevel.Function, "post", Route = "packages")] HttpRequest req,
            ILogger log)
        {
            try
            {
                string requestBody = await new StreamReader(req.Body).ReadToEndAsync();
                var uploadRequest = JsonConvert.DeserializeObject<PackageUploadRequest>(requestBody);

                // Validate request
                if (string.IsNullOrEmpty(uploadRequest.Name) || 
                    string.IsNullOrEmpty(uploadRequest.Version) || 
                    string.IsNullOrEmpty(uploadRequest.Data))
                {
                    return new BadRequestObjectResult(new { error = "Missing required fields" });
                }

                // Validate package
                var validationResult = ValidatePackage(uploadRequest.Name, uploadRequest.Version, uploadRequest.Data);
                if (!validationResult.IsValid)
                {
                    return new BadRequestObjectResult(new { error = validationResult.Message });
                }

                string packageId = $"{uploadRequest.Name}-{uploadRequest.Version}";
                string blobName = $"packages/{uploadRequest.Name}/{uploadRequest.Version}/package.tsk";

                // Check if package already exists
                if (await PackageExistsAsync(uploadRequest.Name, uploadRequest.Version))
                {
                    return new ConflictObjectResult(new { error = "Package version already exists" });
                }

                // Upload to Blob Storage
                var containerClient = _blobServiceClient.GetBlobContainerClient("packages");
                var blobClient = containerClient.GetBlobClient(blobName);

                var packageBytes = Convert.FromBase64String(uploadRequest.Data);
                var blobHttpHeaders = new BlobHttpHeaders
                {
                    ContentType = "application/octet-stream"
                };

                var metadata = new Dictionary<string, string>
                {
                    { "package-name", uploadRequest.Name },
                    { "package-version", uploadRequest.Version },
                    { "upload-time", DateTime.UtcNow.ToString("O") }
                };

                await blobClient.UploadAsync(new MemoryStream(packageBytes), blobHttpHeaders, metadata);

                // Store metadata in Table Storage
                var packageEntity = new PackageEntity
                {
                    PartitionKey = uploadRequest.Name,
                    RowKey = uploadRequest.Version,
                    PackageId = packageId,
                    Name = uploadRequest.Name,
                    Version = uploadRequest.Version,
                    BlobName = blobName,
                    Size = packageBytes.Length,
                    Metadata = JsonConvert.SerializeObject(uploadRequest.Metadata ?? new Dictionary<string, object>()),
                    UploadTime = DateTime.UtcNow,
                    DownloadCount = 0,
                    LastDownload = null
                };

                await _tableClient.AddEntityAsync(packageEntity);

                return new OkObjectResult(new
                {
                    package_id = packageId,
                    download_url = $"https://tusklangcdn.azureedge.net/{blobName}",
                    message = "Package uploaded successfully"
                });
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Upload package error");
                return new ObjectResult(new { error = $"Upload failed: {ex.Message}" }) { StatusCode = 500 };
            }
        }

        [FunctionName("DownloadPackage")]
        public async Task<IActionResult> DownloadPackage(
            [HttpTrigger(AuthorizationLevel.Anonymous, "get", Route = "packages/{name}/{version}/download")] HttpRequest req,
            string name,
            string version,
            ILogger log)
        {
            try
            {
                if (string.IsNullOrEmpty(name) || string.IsNullOrEmpty(version))
                {
                    return new BadRequestObjectResult(new { error = "Missing package name or version" });
                }

                string packageId = $"{name}-{version}";

                // Get package metadata
                var packageEntity = await GetPackageEntityAsync(name, version);
                if (packageEntity == null)
                {
                    return new NotFoundObjectResult(new { error = "Package not found" });
                }

                // Generate SAS token for download
                var containerClient = _blobServiceClient.GetBlobContainerClient("packages");
                var blobClient = containerClient.GetBlobClient(packageEntity.BlobName);

                var sasBuilder = new BlobSasBuilder
                {
                    BlobContainerName = "packages",
                    BlobName = packageEntity.BlobName,
                    Resource = "b",
                    StartsOn = DateTimeOffset.UtcNow,
                    ExpiresOn = DateTimeOffset.UtcNow.AddHours(1)
                };
                sasBuilder.SetPermissions(BlobSasPermissions.Read);

                var sasToken = sasBuilder.ToSasQueryParameters(new Azure.Storage.StorageSharedKeyCredential(
                    Environment.GetEnvironmentVariable("StorageAccountName"),
                    Environment.GetEnvironmentVariable("StorageAccountKey"))).ToString();

                var downloadUrl = $"{blobClient.Uri}?{sasToken}";

                // Update download statistics
                await UpdateDownloadStatsAsync(name, version);

                return new OkObjectResult(new
                {
                    package_id = packageId,
                    download_url = downloadUrl,
                    metadata = JsonConvert.DeserializeObject<Dictionary<string, object>>(packageEntity.Metadata),
                    size = packageEntity.Size
                });
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Download package error");
                return new ObjectResult(new { error = $"Download failed: {ex.Message}" }) { StatusCode = 500 };
            }
        }

        [FunctionName("SearchPackages")]
        public async Task<IActionResult> SearchPackages(
            [HttpTrigger(AuthorizationLevel.Anonymous, "get", Route = "packages/search")] HttpRequest req,
            ILogger log)
        {
            try
            {
                string searchTerm = req.Query["q"].ToString();
                int limit = int.TryParse(req.Query["limit"].ToString(), out int l) ? l : 20;
                int offset = int.TryParse(req.Query["offset"].ToString(), out int o) ? o : 0;

                // Query Table Storage
                var query = _tableClient.QueryAsync<PackageEntity>();
                var packages = new List<PackageEntity>();

                await foreach (var package in query)
                {
                    if (string.IsNullOrEmpty(searchTerm) || 
                        package.Name.Contains(searchTerm, StringComparison.OrdinalIgnoreCase) ||
                        package.Version.Contains(searchTerm, StringComparison.OrdinalIgnoreCase))
                    {
                        packages.Add(package);
                    }
                }

                // Apply pagination
                var results = packages.Skip(offset).Take(limit).Select(p => new
                {
                    package_id = p.PackageId,
                    name = p.Name,
                    version = p.Version,
                    download_count = p.DownloadCount,
                    upload_time = p.UploadTime.ToString("O"),
                    metadata = JsonConvert.DeserializeObject<Dictionary<string, object>>(p.Metadata)
                }).ToList();

                return new OkObjectResult(new
                {
                    packages = results,
                    total = results.Count,
                    next_offset = offset + limit < packages.Count ? offset + limit : (int?)null
                });
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Search packages error");
                return new ObjectResult(new { error = $"Search failed: {ex.Message}" }) { StatusCode = 500 };
            }
        }

        [FunctionName("GetPackageInfo")]
        public async Task<IActionResult> GetPackageInfo(
            [HttpTrigger(AuthorizationLevel.Anonymous, "get", Route = "packages/{name}/{version}")] HttpRequest req,
            string name,
            string version,
            ILogger log)
        {
            try
            {
                if (string.IsNullOrEmpty(name) || string.IsNullOrEmpty(version))
                {
                    return new BadRequestObjectResult(new { error = "Missing package name or version" });
                }

                var packageEntity = await GetPackageEntityAsync(name, version);
                if (packageEntity == null)
                {
                    return new NotFoundObjectResult(new { error = "Package not found" });
                }

                var downloadUrl = $"https://tusklangcdn.azureedge.net/{packageEntity.BlobName}";

                return new OkObjectResult(new
                {
                    package_id = packageEntity.PackageId,
                    name = packageEntity.Name,
                    version = packageEntity.Version,
                    download_url = downloadUrl,
                    size = packageEntity.Size,
                    download_count = packageEntity.DownloadCount,
                    upload_time = packageEntity.UploadTime.ToString("O"),
                    last_download = packageEntity.LastDownload?.ToString("O"),
                    metadata = JsonConvert.DeserializeObject<Dictionary<string, object>>(packageEntity.Metadata)
                });
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Get package info error");
                return new ObjectResult(new { error = $"Failed to get package info: {ex.Message}" }) { StatusCode = 500 };
            }
        }

        private ValidationResult ValidatePackage(string name, string version, string data)
        {
            try
            {
                if (string.IsNullOrEmpty(name) || name.Length > 100)
                {
                    return new ValidationResult { IsValid = false, Message = "Invalid package name" };
                }

                if (string.IsNullOrEmpty(version) || version.Length > 50)
                {
                    return new ValidationResult { IsValid = false, Message = "Invalid package version" };
                }

                if (string.IsNullOrEmpty(data))
                {
                    return new ValidationResult { IsValid = false, Message = "Package data is empty" };
                }

                var packageBytes = Convert.FromBase64String(data);
                if (packageBytes.Length > 50 * 1024 * 1024) // 50MB limit
                {
                    return new ValidationResult { IsValid = false, Message = "Package too large" };
                }

                // Basic TuskLang format validation
                var packageText = Encoding.UTF8.GetString(packageBytes);
                if (!packageText.Trim().StartsWith("#") && !packageText.Contains(":"))
                {
                    return new ValidationResult { IsValid = false, Message = "Invalid TuskLang format" };
                }

                return new ValidationResult { IsValid = true, Message = "Package is valid" };
            }
            catch (Exception ex)
            {
                return new ValidationResult { IsValid = false, Message = $"Validation error: {ex.Message}" };
            }
        }

        private async Task<bool> PackageExistsAsync(string name, string version)
        {
            try
            {
                var packageEntity = await GetPackageEntityAsync(name, version);
                return packageEntity != null;
            }
            catch
            {
                return false;
            }
        }

        private async Task<PackageEntity> GetPackageEntityAsync(string name, string version)
        {
            try
            {
                var response = await _tableClient.GetEntityAsync<PackageEntity>(name, version);
                return response.Value;
            }
            catch
            {
                return null;
            }
        }

        private async Task UpdateDownloadStatsAsync(string name, string version)
        {
            try
            {
                var packageEntity = await GetPackageEntityAsync(name, version);
                if (packageEntity != null)
                {
                    packageEntity.DownloadCount++;
                    packageEntity.LastDownload = DateTime.UtcNow;
                    await _tableClient.UpdateEntityAsync(packageEntity, Azure.ETag.All);
                }
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Failed to update download stats");
            }
        }
    }

    public class PackageUploadRequest
    {
        public string Name { get; set; }
        public string Version { get; set; }
        public string Data { get; set; }
        public Dictionary<string, object> Metadata { get; set; }
    }

    public class PackageEntity : Azure.Data.Tables.ITableEntity
    {
        public string PartitionKey { get; set; }
        public string RowKey { get; set; }
        public DateTimeOffset? Timestamp { get; set; }
        public Azure.ETag ETag { get; set; }

        public string PackageId { get; set; }
        public string Name { get; set; }
        public string Version { get; set; }
        public string BlobName { get; set; }
        public int Size { get; set; }
        public string Metadata { get; set; }
        public DateTime UploadTime { get; set; }
        public int DownloadCount { get; set; }
        public DateTime? LastDownload { get; set; }
    }

    public class ValidationResult
    {
        public bool IsValid { get; set; }
        public string Message { get; set; }
    }
} 